test/integration/test_election.c

```diff
@@ -769,3 +769,114 @@ TEST(election, preVoteNoStaleVotes, setUp, tearDown, 0, cluster_3_params)
+    CLUSTER_SET_DISK_LATENCY(1, 10000);
+
+    /* Disconnect server 1 from server 0, so it won't vote for it. */
+    CLUSTER_SATURATE(0, 1);
+    CLUSTER_SATURATE(1, 0);
+
+    CLUSTER_START;
+
+    /* Server 0 wins elections for term 2, with vote from server 2. */
+    CLUSTER_STEP_UNTIL_HAS_LEADER(1500);
+    munit_assert_int(CLUSTER_LEADER, ==, 0);
+
+    /* Demote server 1 to stand-by. */
+    rv = raft_assign(CLUSTER_RAFT(0), &req, 2 /* ID */, RAFT_STANDBY, NULL);
+    munit_assert_int(rv, ==, 0);
+    CLUSTER_STEP_UNTIL_APPLIED(0, 3 /* entry index */, 1000);
+
+    /* Promote server 3 to voter. */
+    rv = raft_assign(CLUSTER_RAFT(0), &req, 4, RAFT_VOTER, NULL);
+    munit_assert_int(rv, ==, 0);
+
+    /* Wait for server 3 to become aware that it's a voter. */
+    CLUSTER_STEP_UNTIL_APPLIED(3, 4 /* entry index */, 1000);
+
+    /* In the meantime server 1 has timeout and has started an (unsuccessful)
+     * election. */
+    munit_assert_int(CLUSTER_STATE(1), ==, RAFT_CANDIDATE);
+
+    /* Reconnect server 1 to server 0, so it will receive up to index 4,
+     * although it won't persist it since it has a high disk latency. */
+    CLUSTER_DESATURATE(0, 1);
+    CLUSTER_DESATURATE(1, 0);
+
+    /* Wait for server 1 to get contacted by server 0, steps down receive
+     * entries from it */
+    CLUSTER_STEP_UNTIL_STATE_IS(1, RAFT_FOLLOWER, 500);
+    munit_assert_int(raft_last_index(CLUSTER_RAFT(1)), ==, 4);
+
+    /* Create a network partition, with server 0 and 3 in one partition and
+     * server 1 and 2 in another partition. */
+    CLUSTER_SATURATE(0, 1);
+    CLUSTER_SATURATE(1, 0);
+    CLUSTER_SATURATE(0, 2);
+    CLUSTER_SATURATE(2, 0);
+    CLUSTER_SATURATE(3, 1);
+    CLUSTER_SATURATE(1, 3);
+    CLUSTER_SATURATE(3, 2);
+    CLUSTER_SATURATE(2, 3);
+
+    /* Eventually both server 1 and server 2 time out and start elections,
+     * because they have been disconnected from the leader.
+     *
+     * Server 1 is not a voter in the latest configuration at index 4, but it
+     * nevertheless converts to candidate as it's still using the original
+     * configuration at index 3, because it did receive the configuration at
+     * index 4, but hasn't persisted it yet. */
+    CLUSTER_STEP_UNTIL_STATE_IS(1, RAFT_CANDIDATE, 1500);
+    CLUSTER_STEP_UNTIL_STATE_IS(2, RAFT_CANDIDATE, 1500);
+
+    /* Server 2 can't win the election, because it does not consider server 1 a
+     * voter, according to the configuration at index 4.
+     *
+     * Server 1 also can't win the election, because the last index it sends is
+     * the index of its last persisted entry (entry 1), and so server 2 doesn't
+     * grant its vote. */
+    CLUSTER_STEP_UNTIL_ELAPSED(3000);
+    CLUSTER_STEP_UNTIL_STATE_IS(1, RAFT_CANDIDATE, 1500);
+    CLUSTER_STEP_UNTIL_STATE_IS(2, RAFT_CANDIDATE, 1500);
+
+    /* Server 0 is still leader, since it can contact server 3. */
+    munit_assert_int(CLUSTER_STATE(0), ==, RAFT_LEADER);
```

Comment:
Not entirely sure what you mean, the test has 4 servers, indexed from `0` to `3`, with IDs from `1` to `4` (the off-by-one difference between indexes and IDs is a unfortunate, I'd like to fix that eventually, but it's the current convention).

Reply:
You can have any number of servers, with any role combination you wish. An even number of (voting) servers is not recommended because having 4 servers (all voters) doesn't make the cluster more "available" than having 3, since with 4 servers to have a majority you need 3 votes, so you can afford to lose at most 1 server, exactly like when you have 3 servers.

Note that the test sets up 4 servers, but only 3 of them are meant to be voting at any given time (the test mimics automatic roles management in that regard, by demoting an offline voter node and promoting a standby to replace it).